accounting-service/app/src/main/java/com/microservices/accounting/service/PaymentServiceImpl.java

```diff
@@ -0,0 +1,82 @@
+import com.microservices.accounting.temporaryclasses.CardInfo;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.Objects;
+import java.util.Random;
+
+@Service
+public class PaymentServiceImpl implements PaymentService {
+    private final Logger logger = LoggerFactory.getLogger(PaymentServiceImpl.class);
+
+    private PaymentRepository paymentRepository;
+
+    @Autowired
+    public PaymentServiceImpl(PaymentRepository paymentRepository) {
+        this.paymentRepository = paymentRepository;
+    }
+
+    @Override
+    public PaymentDetailsDto invokePayment(InvokePaymentDto invokePaymentDto) {
+        Objects.requireNonNull(invokePaymentDto);
+
+        logger.info("Invoking payment: {}...", invokePaymentDto);
+        CardInfo cardInfo = invokePaymentDto.getUser().getCardInfo();
+        PaymentStatus paymentStatus = (cardInfo == CardInfo.VALID && isRequestApproved())
+                ? PaymentStatus.ACCEPTED
+                : PaymentStatus.DENIED;
+
+        PaymentEntity savedPayment = paymentRepository.save(new PaymentEntity(invokePaymentDto, paymentStatus));
+        logger.info("Payment completed: {}", savedPayment);
+        return new PaymentDetailsDto(savedPayment);
+    }
+
+    @Override
+    @Transactional
+    public PaymentDetailsDto revertPayment(int paymentId) {
+        logger.info("Reverting payment with id: {}", paymentId);
+
+        PaymentEntity paymentToRevert = paymentRepository.findById(paymentId)
+                .orElseThrow(() -> new IllegalArgumentException("No payment with id " + paymentId));
+
+        switch (paymentToRevert.getPaymentStatus()) {
+            case REVERTED:
+                logger.info("Payment with id {} is already reverted", paymentToRevert.getPaymentId());
+                return new PaymentDetailsDto(paymentToRevert);
+            case DENIED:
+                logger.error("Payment with id {} cannot be reverted. Has a denied status", paymentToRevert.getPaymentId());
+                throw new IllegalArgumentException("Denied payment cannot be reverted");
+            case ACCEPTED:
+                if (isRequestApproved()) {
+                    paymentToRevert.setPaymentStatus(PaymentStatus.REVERTED);
+                    PaymentEntity updatedPaymentEntity = paymentRepository.save(paymentToRevert);
+                    logger.info("Payment with id {} is reverted");
+                    return new PaymentDetailsDto(updatedPaymentEntity);
+                } else {
+                    logger.error("Payment reversion cannot be processed");
+                    throw new IllegalStateException("Payment reversion cannot be processed now. Try later");
+                }
+            default:
+                throw new IllegalArgumentException("Unknown payment status");
```

Comment:
State! This is really illegal state of an application, isn't it?

Reply:
IllegalSTATEException

Reply:
I have updated it already. The snippet is outdated